### PR TITLE
FIx race condition in LazyOptional

### PR DESCRIPTION
**Pull request by malte0811**
See #7610 

Also adds a test for both corner cases discussed in the issue (parallel access and contract-violating suppliers).

### CODE REVIEW

File: src/main/java/net/minecraftforge/common/util/LazyOptional.java [hunk 1]
```diff
@@ -23,12 +23,13 @@
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
-import java.util.concurrent.atomic.AtomicReference;
```
Comment (TheCurle) [inline src/main/java/net/minecraftforge/common/util/LazyOptional.java:26]: What's the purpose of moving away from the standard library functions?
The Apache tools are much less useful for this, as we wanted to keep it Atomic, which Apache's Mutable has notable shortcomings with...

Comment (malte0811) [reply src/main/java/net/minecraftforge/common/util/LazyOptional.java:26]: What use would keeping it atomic have in the new implementation? It would most likely be slower (no idea how much), and the multithreading is correct without it (the lock deals with synchronization).

Comment (LexManos) [reply src/main/java/net/minecraftforge/common/util/LazyOptional.java:26]: As I said on the issue, keep the atomic and don't make weird random state value things.
Any claims of it being 'slower' need to actually have statistics to back it up.
Currently your doing a 36 liner change for something that could be 2.

**Comment by LexManos** [comment]
Anyways im tired of arguing with you when i've explicitly told you how to do it.
However this isn't that important of a change do I don't care.
Next time if you want to submit fixes for minor things, don't include random-redesigns along side it.
